### DISCUSSION TITLE
Python API: Get time domain signals after simulation

### DISCUSSION THREAD

**Discussion by VolkerMuehlhaus** [thliebig/openEMS-Project#156] [Q&A Usage] [answered]
Hi all,

I am using openEMS with **Python** interface and wonder if there is an easy way to access the **time domain** results? 
In the **Matlab** interface this is documented in the code, but for Python API I don't find something similar!?

```
% output: 
%   % output signals/values in time domain (TD):
%   port.ut.tot     total voltage (time-domain)
%   port.ut.time    voltage time vector
%   port.it.tot     total current (time-domain)
%   port.it.time    current time vector
```

The data is stored to output files port_ut_1 and port_it_1 (for single port case), but I haven't found a way to access this from the API directly. Do I really need to read the files from disk when using Python API?

Best regards
Volker

**Comment by LubomirJagos** [comment, marked as answer]
Good way to get similar IDE for python as octave has is to install Spyder and set in Tools > Preferences path to python which are you using to run simulations since it has libraries installed.
https://www.spyder-ide.org/

then you have there variable inspector, in python you can't click through variables but you can then use dir() or add other variables as watch for variable.

What I tried this is access to data:
  - portVariable.u_data.ui_val[0] = U over time
  - portVariable.i_data.ui_val[0] = I over time
  - portVariable.u_data.ui_time[0] = timesteps
  - portVariable.i_data.ui_time[0] = timesteps (should be same as for voltage)
  ...
  
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/7ddfa1ad-524c-4d5f-a0ef-a510dd386482)

**Reply by LubomirJagos** [reply]
I forgot to mention that you have to install some jupyter libraries into python to make it running, it will show you there what to install to got interactive commnad line there

**Reply by VolkerMuehlhaus** [reply]
```
    portVariable.u_data.ui_val[0] = U over time
    portVariable.i_data.ui_val[0] = I over time
    portVariable.u_data.ui_time[0] = timesteps
    portVariable.i_data.ui_time[0] = timesteps (should be same as for voltage)
```

Many thanks Lubomir, this worked! 
I simply used it from my existing Visual Studio Code environment, and will look into Spyder later.

**Reply by VolkerMuehlhaus** [reply]
Hello Lubomir, I have given up on Spyder, the latest default Windows installation reports some SQlite errors when loading the kernel. Would be nice to have the capacibility that you show, but I have no time and patience for troubleshooting that, and keep using my existing python environment. Many thanks again for guiding me with the time domain signals! BR, Volker

**Reply by LubomirJagos** [reply]
hmmm, sorry to hear that, I installed it just recently when was working on python script generator for openEMS and had no problems with it.
Briefly looking Visual Code also has similar capabilities for python. It's easier to discover python interface interactively using some IDE than running it in batch since python has good reflection (you can inspect variables and data on the fly).

There are also some differencies in python interface last one what I hit is that NF2FF function CalcNF2FF requires phi and theta angle as is written in help page https://docs.openems.de/python/openEMS/nf2ff.html but using python these angles are in degrees, in octave interface it's radians, but output of both python and octave results is in radians (due it's running same underlying C++ function).

Have to add it into documentation, there some little things which are easy to observe when using IDE and stepping through code :D
